source/drivers/Sensor.cpp

```diff
@@ -0,0 +1,247 @@
+/*
+The MIT License (MIT)
+
+Copyright (c) 2017 Lancaster University.
+Copyright (c) 2018 Paul ADAM, inidinn.com
+
+Permission is hereby granted, free of charge, to any person obtaining a
+copy of this software and associated documentation files (the "Software"),
+to deal in the Software without restriction, including without limitation
+the rights to use, copy, modify, merge, publish, distribute, sublicense,
+and/or sell copies of the Software, and to permit persons to whom the
+Software is furnished to do so, subject to the following conditions:
+
+The above copyright notice and this permission notice shall be included in
+all copies or substantial portions of the Software.
+
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
+THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+DEALINGS IN THE SOFTWARE.
+*/
+
+/**
+ * Class definition for a generic analog sensor, that takes the general form of a logarithmic response to a sensed value, in a potential divider.
+ * Implements a base class for such a sensor, using the Steinhart-Hart equation to delineate a result.
+ */
+
+#include "CodalConfig.h"
+#include "Sensor.h"
+#include "ErrorNo.h"
+#include "CodalConfig.h"
+#include "Event.h"
+#include "CodalCompat.h"
+#include "CodalFiber.h"
+#include "Timer.h"
+
+using namespace codal;
+
+/**
+ * Constructor.
+ *
+ * Creates a generic Sensor.
+ *
+ * @param id The ID of this compoenent e.g. DEVICE_ID_THERMOMETER
+ */
+Sensor::Sensor(uint16_t id)
+{
+    this->id = id;
+    this->sensitivity = 868;
+
+    // Configure for a 2 Hz update frequency by default.
+    if(EventModel::defaultEventBus)
+        EventModel::defaultEventBus->listen(id, SENSOR_UPDATE_NEEDED, this, &Sensor::onSampleEvent, MESSAGE_BUS_LISTENER_IMMEDIATE);
+
+    setPeriod(500);
+}
+
+/*
+ * Event Handler for periodic sample timer
+ */
+void Sensor::onSampleEvent(Event)
+{
+    updateSample();
+}
+
+/*
+ * Determines the instantaneous value of the sensor, in SI units, and returns it.
+ *
+ * @return The current value of the sensor.
+ */
+int Sensor::getValue()
+{
+    return (int)sensorValue;
+}
+
+/**
+ * Updates the internal reading of the sensor. Typically called periodically.
+ */
+void Sensor::updateSample()
+{
+    uint32_t value = readValue();
+
+    // If this is the first reading performed, take it a a baseline. Otherwise, perform a decay average to smooth out the data.
+    if (!(status & SENSOR_INITIALISED))
+    {
+        sensorValue = (uint16_t)value;
+        status |=  SENSOR_INITIALISED;
+        // set Period to default value
+        setPeriod(500);
```

Comment:
What about something like:

`setPeriod(getPeriod() > 0 ? getPeriod() : 500)`

P.s. 500 should be a #define (probably)